tools/clang/unittests/HLSL/DxilContainerTest.cpp

```diff
@@ -1762,6 +1763,296 @@ TEST_F(DxilContainerTest, DxcUtils_CreateReflection) {
+      // Check that bytes in this struct after offset are still 0xFE
+      uint8_t *pBytes = (uint8_t *)this;
+      for (size_t i = offset; i < sizeof(TestDescStruct); i++) {
+        if (pBytes[i] != 0xFE)
+          return false;
+      }
+      return true;
+    }
+    bool CheckBytesAfterStream() {
+      // Check that bytes in this struct after Stream are still 0xFE
+      return CheckRemainingBytes(offsetof(TestDescStruct, paramDesc.Stream) +
+                                 sizeof(paramDesc.Stream));
+    }
+    bool CheckBytesAfterParamDesc() {
+      // Check that bytes in this struct after paramDesc are still 0xFE
+      return CheckRemainingBytes(offsetof(TestDescStruct, paramDesc) +
+                                 sizeof(paramDesc));
+    }
+  };
+
+  // ID3D12LibraryReflection
+  {
+    CComPtr<IDxcOperationResult> pResult;
+    CComPtr<IDxcResult> pResultV2;
+    CComPtr<IDxcBlob> pReflectionPart;
+    CComPtr<IDxcBlobEncoding> pSource;
+    CreateBlobFromText(Ref1_Shader, &pSource);
+    VERIFY_SUCCEEDED(pCompiler->Compile(pSource, L"hlsl.hlsl", L"", L"lib_6_3",
+                                        nullptr, 0, nullptr, 0, nullptr,
+                                        &pResult));
+    VERIFY_SUCCEEDED(pResult->QueryInterface(&pResultV2));
+    VERIFY_SUCCEEDED(pResultV2->GetOutput(
+        DXC_OUT_REFLECTION, IID_PPV_ARGS(&pReflectionPart), nullptr));
+    DxcBuffer buffer = {pReflectionPart->GetBufferPointer(),
+                        pReflectionPart->GetBufferSize(), 0};
+
+    {
+      CComPtr<ID3D12LibraryReflection> pLibraryReflection;
+      VERIFY_SUCCEEDED(
+          pUtils->CreateReflection(&buffer, IID_PPV_ARGS(&pLibraryReflection)));
+
+      CComPtr<ID3D12LibraryReflection> pLibraryReflection2;
+      VERIFY_SUCCEEDED(pLibraryReflection->QueryInterface(
+          IID_PPV_ARGS(&pLibraryReflection2)));
+      CComPtr<IUnknown> pUnknown;
+      VERIFY_SUCCEEDED(
+          pLibraryReflection->QueryInterface(IID_PPV_ARGS(&pUnknown)));
+      CComPtr<ID3D12ShaderReflection> pShaderReflection;
+      VERIFY_FAILED(
+          pLibraryReflection->QueryInterface(IID_PPV_ARGS(&pShaderReflection)));
+    }
+
+    { // Allow IUnknown
+      CComPtr<IUnknown> pUnknown;
+      VERIFY_SUCCEEDED(
+          pUtils->CreateReflection(&buffer, IID_PPV_ARGS(&pUnknown)));
+      CComPtr<ID3D12LibraryReflection> pLibraryReflection;
+      VERIFY_SUCCEEDED(
+          pUnknown->QueryInterface(IID_PPV_ARGS(&pLibraryReflection)));
+    }
+
+    { // Fail to create with incorrect interface
+      CComPtr<ID3D12ShaderReflection> pShaderReflection;
+      VERIFY_ARE_EQUAL(
+          E_NOINTERFACE,
+          pUtils->CreateReflection(&buffer, IID_PPV_ARGS(&pShaderReflection)));
+    }
+  }
+
+  // Supported shader reflection interfaces defined by legacy APIs
+  const GUID IID_ID3D11ShaderReflection_43 = {
+      0x0a233719,
+      0x3960,
+      0x4578,
+      {0x9d, 0x7c, 0x20, 0x3b, 0x8b, 0x1d, 0x9c, 0xc1}};
```

Comment:
Should these come in from a header, instead of defining them here?

Reply:
They aren't defined in any header I could include, and they aren't versioned the way I need.  Normally, you'd link to one SDK version and the GUID for ID3D11ShaderReflection would be one of these, depending on your SDK version.  I don't think there's any accessible standard way of using that here.  This is the simplest way to test behavior differences that depend on which interface GUID you used to create the reflection.  These GUIDs are similarly defined in DxilContainerReflection.cpp for the same reason.